### PR TITLE
CCS with minimize_roundtrips performs incremental merges of each SearchResponse

### PR DESCRIPTION
**Pull request by quux00**
To help address the issue of slow-to-respond clusters in a cross-cluster search,
async-search based CCS with minimize_roundtrips=true performs incremental
merges of each SearchResponse as they come in from each cluster (including
the local cluster).

This means, any time the user calls `GET _async_search/:id`, they will now get
search hits and/or aggregation results from any clusters that have finished so far.

The `is_running` field in the async-search response should be used to determine
whether at least one cluster has still not reported back.

### CODE REVIEW

**Comment by quux00** [comment]
Example run against my local "blogs" index.

Query:

```
POST blogs,remote*:blogs*/_async_search?ccs_minimize_roundtrips=true&wait_for_completion_timeout=100ms&keep_on_completion=true
{
  "size": 3,
  "query": {
    "error_query": {
      "indices": [
        {
          "name": "*",
          "error_type": "none",
          "stall_time_seconds": 1
        },
        {
          "name": "remote1:*",
          "error_type": "none",
          "stall_time_seconds": 4
        },
        {
          "name": "remote2:*",
          "error_type": "none",
          "stall_time_seconds": 16
        }
      ]
    }
  },
  "aggs": {
    "indexgroup": {
      "terms": {
        "field": "_index"
      }
    }
  }
}
```

<details>
  <summary><b>Response after 2 out of 3 clusters have reported back</b> (toggle to view)</summary>
<pre>
GET _async_search/FmQ1ZlhxcWltUzdLNFNpTUpsLUJJYXccT0l6RmpKMkhSaXFzTDNaMDRGVHdBZzoxMjQwNg
</pre>

<pre>
{
  "id": "FmQ1ZlhxcWltUzdLNFNpTUpsLUJJYXccT0l6RmpKMkhSaXFzTDNaMDRGVHdBZzoxMjQwNg==",
  "is_partial": true,
  "is_running": true,
  "start_time_in_millis": 1700593778293,
  "expiration_time_in_millis": 1701025778293,
  "response": {
    "took": 4042,
    "timed_out": false,
    "num_reduce_phases": 3,
    "_shards": {
      "total": 6,
      "successful": 6,
      "skipped": 0,
      "failed": 0
    },
    "_clusters": {
      "total": 3,
      "successful": 2,
      "skipped": 0,
      "running": 1,
      "partial": 0,
      "failed": 0,
      "details": {
        "(local)": {
          "status": "successful",
          "indices": "blogs",
          "took": 1046,
          "timed_out": false,
          "_shards": {
            "total": 3,
            "successful": 3,
            "skipped": 0,
            "failed": 0
          }
        },
        "remote2": {
          "status": "running",
          "indices": "blogs*",
          "timed_out": false
        },
        "remote1": {
          "status": "successful",
          "indices": "blogs*",
          "took": 4022,
          "timed_out": false,
          "_shards": {
            "total": 3,
            "successful": 3,
            "skipped": 0,
            "failed": 0
          }
        }
      }
    },
    "hits": {
      "total": {
        "value": 2798,
        "relation": "eq"
      },
      "max_score": 1,
      "hits": [
        {
          "_index": "remote1:blogs",
          "_id": "HZLL8osBrgB4eJrgKv27",
          "_score": 1,
          "_source": {
            "content": "produce late bed way network what however lose who enter lay southern hour this enough foot gas much believe involve hotel lay center third specific side treatment author big box far loss need yeah entire worker son measure law much road responsibility computer camera keep employee race into yourself treatment part mission attorney black her test deal range society able air bring technology may travel prevent century man including public crime society nothing why war capital imagine have lot trouble same attack drive adult senior run culture Mr lose eat TV guy pull sound recognize as clear hospital seat eight east act control wrong threat paper much phone by picture choice modern create administration blue station challenge economic stand measure without fish see send doctor scene style raise drive tough bring up provide future prevent red always prepare long environment lawyer follow however",
            "url": "/blog/over-natural",
            "category": [
              "bltfaae4466058cc7d6"
            ],
            "publish_date": "2017-08-12T07:48:25.000Z",
            "authors": [
              {
                "full_name": "Melinda Brady",
                "uid": "e4bf9c68-c5b4-4aac-aa85-8182cb9b4050",
                "company": "Phelps, Shepard and Buchanan",
                "job_title": "Furniture conservator/restorer",
                "last_name": "Mejia",
                "first_name": "David"
              }
            ],
            "tags": {
              "level": [
                "2"
              ],
              "elastic_stack": [],
              "topic": [],
              "use_case": [],
              "use_cases": [],
              "tags": [],
              "product": [
                "filebeat"
              ],
              "industry": [
                "healthcare"
              ]
            },
            "locale": "en-us",
            "title": "final account federal individual understand perhaps different attorney put"
          }
        },
        {
          "_index": "remote1:blogs",
          "_id": "HpLL8osBrgB4eJrgKv27",
          "_score": 1,
          "_source": {
            "content": "room everyone deep nothing maybe contain leader to study together power food pretty particularly top relate provide official walk best collection generation short popular free per ok threat adult spend notice part find write stock space hear customer significant it fly your against generation task student add work term same dream enjoy several west look blue space avoid foot available fund look far parent window term among model pressure whole personal go high should manager single poor or form here lot with ago poor risk",
            "url": "/blog/civil-find",
            "category": [
              "xyzfaae446605123456"
            ],
            "publish_date": "2018-08-11T21:35:29.000Z",
            "authors": [
              {
                "full_name": "Steven Day",
                "uid": "1fb3225f-3dd1-4af8-8af8-239bec183317",
                "company": "Herman-Herring",
                "job_title": "Legal executive",
                "last_name": "West",
                "first_name": "William"
              }
            ],
            "tags": {
              "level": [
                "3"
              ],
              "elastic_stack": [],
              "topic": [],
              "use_case": [],
              "use_cases": [],
              "tags": [],
              "product": [
                "elasticsearch"
              ],
              "industry": [
                "healthcare"
              ]
            },
            "locale": "en-us",
            "title": "young chance popular mention still"
          }
        },
        {
          "_index": "remote1:blogs",
          "_id": "IpLL8osBrgB4eJrgKv27",
          "_score": 1,
          "_source": {
            "content": "word government unit clear number west already light realize suggest low rise adult against much own season police shoulder change effect public garden reality best world program beautiful someone",
            "url": "/blog/power-performance",
            "category": [
              "abcdef6655058cc7d6"
            ],
            "publish_date": "2009-11-04T04:08:07.000Z",
            "authors": [
              {
                "full_name": "Brian Wilson",
                "uid": "52c6253d-e743-4a01-b4de-f73be7f66c65",
                "company": "Jackson, Moreno and Brown",
                "job_title": "Race relations officer",
                "last_name": "Vaughn",
                "first_name": "Tammy"
              }
            ],
            "tags": {
              "level": [
                "1"
              ],
              "elastic_stack": [],
              "topic": [],
              "use_case": [],
              "use_cases": [],
              "tags": [],
              "product": [
                "es-hadoop"
              ],
              "industry": [
                "retail"
              ]
            },
            "locale": "en-us",
            "title": "oil left success enough their require"
          }
        }
      ]
    },
    "aggregations": {
      "indexgroup": {
        "doc_count_error_upper_bound": 0,
        "sum_other_doc_count": 0,
        "buckets": [
          {
            "key": "blogs",
            "doc_count": 1399
          },
          {
            "key": "remote1:blogs",
            "doc_count": 1399
          }
        ]
      }
    }
  }
}
</pre>
</details>


<details>
  <summary><b>Response after all clusters have reported back</b> (toggle to view)</summary>
<pre>
GET _async_search/FmQ1ZlhxcWltUzdLNFNpTUpsLUJJYXccT0l6RmpKMkhSaXFzTDNaMDRGVHdBZzoxMjQwNg
</pre>

<pre>
{
  "id": "FmQ1ZlhxcWltUzdLNFNpTUpsLUJJYXccT0l6RmpKMkhSaXFzTDNaMDRGVHdBZzoxMjQwNg==",
  "is_partial": false,
  "is_running": false,
  "start_time_in_millis": 1700593778293,
  "expiration_time_in_millis": 1701025778293,
  "completion_time_in_millis": 1700593794334,
  "response": {
    "took": 16041,
    "timed_out": false,
    "num_reduce_phases": 4,
    "_shards": {
      "total": 9,
      "successful": 9,
      "skipped": 0,
      "failed": 0
    },
    "_clusters": {
      "total": 3,
      "successful": 3,
      "skipped": 0,
      "running": 0,
      "partial": 0,
      "failed": 0,
      "details": {
        "(local)": {
          "status": "successful",
          "indices": "blogs",
          "took": 1046,
          "timed_out": false,
          "_shards": {
            "total": 3,
            "successful": 3,
            "skipped": 0,
            "failed": 0
          }
        },
        "remote2": {
          "status": "successful",
          "indices": "blogs*",
          "took": 16027,
          "timed_out": false,
          "_shards": {
            "total": 3,
            "successful": 3,
            "skipped": 0,
            "failed": 0
          }
        },
        "remote1": {
          "status": "successful",
          "indices": "blogs*",
          "took": 4022,
          "timed_out": false,
          "_shards": {
            "total": 3,
            "successful": 3,
            "skipped": 0,
            "failed": 0
          }
        }
      }
    },
    "hits": {
      "total": {
        "value": 4197,
        "relation": "eq"
      },
      "max_score": 1,
      "hits": [
        {
          "_index": "remote1:blogs",
          "_id": "HZLL8osBrgB4eJrgKv27",
          "_score": 1,
          "_source": {
            "content": "produce late bed way network what however lose who enter lay southern hour this enough foot gas much believe involve hotel lay center third specific side treatment author big box far loss need yeah entire worker son measure law much road responsibility computer camera keep employee race into yourself treatment part mission attorney black her test deal range society able air bring technology may travel prevent century man including public crime society nothing why war capital imagine have lot trouble same attack drive adult senior run culture Mr lose eat TV guy pull sound recognize as clear hospital seat eight east act control wrong threat paper much phone by picture choice modern create administration blue station challenge economic stand measure without fish see send doctor scene style raise drive tough bring up provide future prevent red always prepare long environment lawyer follow however",
            "url": "/blog/over-natural",
            "category": [
              "bltfaae4466058cc7d6"
            ],
            "publish_date": "2017-08-12T07:48:25.000Z",
            "authors": [
              {
                "full_name": "Melinda Brady",
                "uid": "e4bf9c68-c5b4-4aac-aa85-8182cb9b4050",
                "company": "Phelps, Shepard and Buchanan",
                "job_title": "Furniture conservator/restorer",
                "last_name": "Mejia",
                "first_name": "David"
              }
            ],
            "tags": {
              "level": [
                "2"
              ],
              "elastic_stack": [],
              "topic": [],
              "use_case": [],
              "use_cases": [],
              "tags": [],
              "product": [
                "filebeat"
              ],
              "industry": [
                "healthcare"
              ]
            },
            "locale": "en-us",
            "title": "final account federal individual understand perhaps different attorney put"
          }
        },
        {
          "_index": "remote1:blogs",
          "_id": "HpLL8osBrgB4eJrgKv27",
          "_score": 1,
          "_source": {
            "content": "room everyone deep nothing maybe contain leader to study together power food pretty particularly top relate provide official walk best collection generation short popular free per ok threat adult spend notice part find write stock space hear customer significant it fly your against generation task student add work term same dream enjoy several west look blue space avoid foot available fund look far parent window term among model pressure whole personal go high should manager single poor or form here lot with ago poor risk",
            "url": "/blog/civil-find",
            "category": [
              "xyzfaae446605123456"
            ],
            "publish_date": "2018-08-11T21:35:29.000Z",
            "authors": [
              {
                "full_name": "Steven Day",
                "uid": "1fb3225f-3dd1-4af8-8af8-239bec183317",
                "company": "Herman-Herring",
                "job_title": "Legal executive",
                "last_name": "West",
                "first_name": "William"
              }
            ],
            "tags": {
              "level": [
                "3"
              ],
              "elastic_stack": [],
              "topic": [],
              "use_case": [],
              "use_cases": [],
              "tags": [],
              "product": [
                "elasticsearch"
              ],
              "industry": [
                "healthcare"
              ]
            },
            "locale": "en-us",
            "title": "young chance popular mention still"
          }
        },
        {
          "_index": "remote1:blogs",
          "_id": "IpLL8osBrgB4eJrgKv27",
          "_score": 1,
          "_source": {
            "content": "word government unit clear number west already light realize suggest low rise adult against much own season police shoulder change effect public garden reality best world program beautiful someone",
            "url": "/blog/power-performance",
            "category": [
              "abcdef6655058cc7d6"
            ],
            "publish_date": "2009-11-04T04:08:07.000Z",
            "authors": [
              {
                "full_name": "Brian Wilson",
                "uid": "52c6253d-e743-4a01-b4de-f73be7f66c65",
                "company": "Jackson, Moreno and Brown",
                "job_title": "Race relations officer",
                "last_name": "Vaughn",
                "first_name": "Tammy"
              }
            ],
            "tags": {
              "level": [
                "1"
              ],
              "elastic_stack": [],
              "topic": [],
              "use_case": [],
              "use_cases": [],
              "tags": [],
              "product": [
                "es-hadoop"
              ],
              "industry": [
                "retail"
              ]
            },
            "locale": "en-us",
            "title": "oil left success enough their require"
          }
        }
      ]
    },
    "aggregations": {
      "indexgroup": {
        "doc_count_error_upper_bound": 0,
        "sum_other_doc_count": 0,
        "buckets": [
          {
            "key": "blogs",
            "doc_count": 1399
          },
          {
            "key": "remote1:blogs",
            "doc_count": 1399
          },
          {
            "key": "remote2:blogs",
            "doc_count": 1399
          }
        ]
      }
    }
  }
}
</pre>
</details>

**Comment by quux00** [comment]
@elasticsearchmachine run elasticsearch-ci/ubuntu-2004/destructiveDistroTest.archives/packaging-tests-unix-sample

File: server/src/main/java/org/elasticsearch/action/search/SearchResponseMerger.java
Comment (quux00) [inline server/src/main/java/org/elasticsearch/action/search/SearchResponseMerger.java:136]: Now that getMergedResponse can be called from multiple threads (not just the callback thread of the final cluster to report back), I've synchonized this method and the getMergedResponse method to avoid consistency errors.

**Comment by quux00** [comment]
@elasticmachine test this please

**Review by piergm — APPROVED** [review]
LGTM! Nice one @quux00 👍

**Comment by quux00** [comment]
One change I've noticed in my testing. While the search is still running (not all clusters have reported back), the `took` value no longer updates to now() - startTime. It stays at the last `took` value of the last merge (when `SearchResponseMerger.add` was called back a ccsRemoteReduce callback thread).

Example:
```
{
  "id": "FmdCNkhZVGhHUWFPUENTTng3OUhLUFEcOGZXZE9oNEhSOW1NX2xwMGV3Ti15dzo2MTY3MA==",
  "is_partial": true,
  "is_running": true,
  "start_time_in_millis": 1700675356605,
  "expiration_time_in_millis": 1701107356605,
  "response": {
    "took": 12,    // this value won't change until another cluster merges in - it used to update to now()-start_time
    "timed_out": false,
    "num_reduce_phases": 3,
    "_shards": {
      "total": 12,
      "successful": 12,
      "skipped": 2,
      "failed": 0
    },
    "_clusters": {
      "total": 3,
      "successful": 2,
      "skipped": 0,
      "running": 1,
      "partial": 0,
      "failed": 0,
      "details": {
        "(local)": {
          "status": "successful",
          "indices": "*",
          "took": 9,
          "timed_out": false,
          "_shards": {
            "total": 7,
            "successful": 7,
            "skipped": 2,
            "failed": 0
          }
        },
        "remote2": {
          "status": "running",
          "indices": "*",
          "timed_out": false
        },
        "remote1": {
          "status": "successful",
          "indices": "*",
          "took": 11,
          "timed_out": false,
          "_shards": {
            "total": 5,
            "successful": 5,
            "skipped": 0,
            "failed": 0
          }
        }
      }
    },
...
```

I can add in changes to preserve this original behavior, but it means copying the last merged SearchResponse to a new SearchResponse each time the `GET _async_search/:id` endpoint is called. Should I do that?

**Comment by quux00** [comment]
@elasticsearchmachine run elasticsearch-ci/docs

File: server/src/main/java/org/elasticsearch/action/search/SearchResponseMerger.java
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/search/SearchResponseMerger.java:145]: It feels a bit weird to carry the progress listener in SearchResponseMerger. Would it be an option to call nofityCcsReduce after we call SearchResponseMerger#add ?

File: server/src/main/java/org/elasticsearch/action/search/SearchResponseMerger.java
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/search/SearchResponseMerger.java:156]: see comment above on the need for synchronized

**Review by javanna — COMMENTED** [review]
I left a couple of minor comments, this looks good though. One additional question I have is whether exposing incremental hits results should be done more broadly (and not only for CCS), and as a separate change. I would be happy with this change exposing results as they come in from remote clusters, but only for aggs.

File: server/src/main/java/org/elasticsearch/action/search/SearchResponseMerger.java
Comment (quux00) [reply server/src/main/java/org/elasticsearch/action/search/SearchResponseMerger.java]: I reviewed the code more carefully and everything it calls (such as SearchProgressListener.ccsReduce). Without synchronization here everything it does and calls must be thread safe. I believe it is (as the AsyncTask.SearchProgressListener.ccsReduce call to MutableSearchThread.updatePartialResponse is `synchronized`.

So I believe the synchronization here can be removed (and on `add`) but it is a little more risky that any non-thread safe changes in the future will cause issues. I added code comments indicating that these two methods must remain thread-safe.

There are also tests already in SearchResponseTests that test thread safety. 

The one thing this "breaks" or interferes with is that that optimization to not do an incremental merge on the last SearchResponse sent to the `add` method. Say there are 2 responses still to come in and both call add at the same time. In that case, in the tests I'm getting non-deterministic results around how many incremental merges happened, since there is a (benign) race condition. I've adjusted the tests to only run asserts against the number of incremental merges if the test runs serially rather than concurrently. I ran the tests 100,000 times and they did not fail.

However, given the non-deterministic nature of the optimization now, requiring extra work in the tests and being initiall confusing to code readers/maintainers, I wonder if we should remove this optimization.

**Comment by javanna** [comment]
@quux00 I mean that it would be enough in my opinion to have mrt=true to expose incremental aggs results. it would feel more natural to expose incremental top hits results as a whole instead of only for mrt=true (hence CCS). What do you think?

File: server/src/main/java/org/elasticsearch/action/search/SearchResponseMerger.java
Comment (javanna) [reply server/src/main/java/org/elasticsearch/action/search/SearchResponseMerger.java:150]: It's  not a huge deal on my end, but I think that SearchResponseMerger is in principle a lightweight class that is able to merge search responses together, without knowing too much about the context in which it does that. Making it aware of progress listener makes it more complex, especially with the null/noop check. I find that calling notifyCCSReduce outside of it would make sure that we call that only when needed and removes the need for the check. I would not feel bad about CCSACtionListener becoming aware of SearchProgressListener instead. It feels actually more natural to have a listener notify another listener, compared to a sync component notifying a listener of some of its actions.

Comment (javanna) [reply server/src/main/java/org/elasticsearch/action/search/SearchResponseMerger.java:150]: To add to that, I did not realize that `add` calls` getMergedResponse`. I focused on the conditional and not on what it does within the if block :)

This looks like bad practice to me. I think that add should be called from a consumer, and the same as getMergedResponse.

**Review by javanna — COMMENTED** [review]
I did another round, I suppose it's ok to delay working on #62935 , although odd that partial top hits results are only exposed when CCS with minimized roundtrips. It is also true that we have a clear scenario where this is useful, and that involves specifically CCS and avoiding to wait for slower clusters.

File: server/src/main/java/org/elasticsearch/action/search/SearchResponseMerger.java
Comment (quux00) [reply server/src/main/java/org/elasticsearch/action/search/SearchResponseMerger.java:150]: OK. I will close this PR and start a new one.

**Comment by quux00** [comment]
Closing this PR since we want to change the implementation around how/where getMergedResponse is called and that there are merge conflicts around decRef/incRef of SearchResponse and SearchResponseMergers. It will be easier to just start over once those PRs are merged.

**Pull request closed without merging**
